File: sdks/python/hyperspace/math.py
from typing import Sequence, List
import math


def _dot(a: Sequence[float], b: Sequence[float]) -> float:
    return sum(x * y for x, y in zip(a, b))


def _norm_sq(v: Sequence[float]) -> float:
    return _dot(v, v)


def _project_to_ball(x: Sequence[float], c: float) -> List[float]:
    n = math.sqrt(max(_norm_sq(x), 0.0))
    max_n = (1.0 / math.sqrt(c)) - 1e-9
    if n <= max_n or n <= 1e-15:
        return list(x)
    s = max_n / n
    return [s * v for v in x]


def mobius_add(x: Sequence[float], y: Sequence[float], c: float = 1.0) -> List[float]:
    if len(x) != len(y):
        raise ValueError("Dimension mismatch")
    if c <= 0.0:
        raise ValueError("Curvature c must be > 0")
    xy = _dot(x, y)
    x2 = _norm_sq(x)
    y2 = _norm_sq(y)
    num_left = 1.0 + 2.0 * c * xy + c * y2
    num_right = 1.0 - c * x2
    den = 1.0 + 2.0 * c * xy + (c * c) * x2 * y2
    if abs(den) < 1e-15:
        raise ValueError("Möbius addition denominator is too close to zero")
    return [(num_left * xi + num_right * yi) / den for xi, yi in zip(x, y)]


def exp_map(x: Sequence[float], v: Sequence[float], c: float = 1.0) -> List[float]:
    if len(x) != len(v):
        raise ValueError("Dimension mismatch")
    if c <= 0.0:
        raise ValueError("Curvature c must be > 0")
    x2 = _norm_sq(x)
    v_norm = math.sqrt(max(_norm_sq(v), 0.0))
    if v_norm < 1e-15:
        return list(x)
    lambda_x = 2.0 / max(1.0 - c * x2, 1e-15)
    scale = math.tanh(math.sqrt(c) * lambda_x * v_norm / 2.0) / (math.sqrt(c) * v_norm)
    step = [scale * vi for vi in v]
    return mobius_add(x, step, c=c)


def log_map(x: Sequence[float], y: Sequence[float], c: float = 1.0) -> List[float]:
    if len(x) != len(y):
        raise ValueError("Dimension mismatch")
    if c <= 0.0:
        raise ValueError("Curvature c must be > 0")
    neg_x = [-xi for xi in x]
    delta = mobius_add(neg_x, y, c=c)
    delta_norm = math.sqrt(max(_norm_sq(delta), 0.0))
    if delta_norm < 1e-15:
        return [0.0 for _ in x]
    x2 = _norm_sq(x)
    lambda_x = 2.0 / max(1.0 - c * x2, 1e-15)
    factor = (2.0 / (lambda_x * math.sqrt(c))) * math.atanh(
        min(math.sqrt(c) * delta_norm, 1.0 - 1e-15)
    )
    return [factor * di / delta_norm for di in delta]


def riemannian_gradient(x: Sequence[float], euclidean_grad: Sequence[float], c: float = 1.0) -> List[float]:
    if len(x) != len(euclidean_grad):
        raise ValueError("Dimension mismatch")
    if c <= 0.0:
        raise ValueError("Curvature c must be > 0")
    x2 = _norm_sq(x)
    lambda_x = 2.0 / max(1.0 - c * x2, 1e-15)
    scale = 1.0 / (lambda_x * lambda_x)
    return [scale * g for g in euclidean_grad]


def _gyro(u: Sequence[float], v: Sequence[float], w: Sequence[float], c: float = 1.0) -> List[float]:
    uv = mobius_add(u, v, c=c)
    vw = mobius_add(v, w, c=c)
    left = mobius_add(u, vw, c=c)
    neg_uv = [-z for z in uv]
    return mobius_add(neg_uv, left, c=c)


def parallel_transport(x: Sequence[float], y: Sequence[float], v: Sequence[float], c: float = 1.0) -> List[float]:
    if len(x) != len(y) or len(x) != len(v):
        raise ValueError("Dimension mismatch")
    if c <= 0.0:
        raise ValueError("Curvature c must be > 0")
    neg_x = [-xi for xi in x]
    gyr = _gyro(y, neg_x, v, c=c)
    lambda_x = 2.0 / max(1.0 - c * _norm_sq(x), 1e-15)
    lambda_y = 2.0 / max(1.0 - c * _norm_sq(y), 1e-15)
    scale = lambda_x / lambda_y
    return [scale * gi for gi in gyr]


def frechet_mean(points: Sequence[Sequence[float]], c: float = 1.0, max_iter: int = 64, tol: float = 1e-8) -> List[float]:
    if not points:
        raise ValueError("Points set cannot be empty")
    if c <= 0.0:
        raise ValueError("Curvature c must be > 0")
    dim = len(points[0])
    if any(len(p) != dim for p in points):
        raise ValueError("Dimension mismatch")
    mu = _project_to_ball(points[0], c)
    for _ in range(max(1, max_iter)):
        grad = [0.0] * dim
        for p in points:
            lg = log_map(mu, p, c=c)
            for i in range(dim):
                grad[i] += lg[i]
        inv = 1.0 / float(len(points))
        grad = [g * inv for g in grad]
        g_norm = math.sqrt(max(_norm_sq(grad), 0.0))
        mu = exp_map(mu, grad, c=c)
        mu = _project_to_ball(mu, c)
    return mu


# ==========================================
# Lorentz Model Math (Hyperboloid)
# ==========================================

def lorentz_product(u: Sequence[float], v: Sequence[float]) -> float:
    """Computes the Minkowski inner product (Lorentz product) between two vectors."""
    if not u or not v:
        return 0.0
    dot = -u[0] * v[0]
    for ui, vi in zip(u[1:], v[1:]):
        dot += ui * vi
    return dot


def lorentz_dist(u: Sequence[float], v: Sequence[float]) -> float:
    """Computes the Lorentz distance between two points on the hyperboloid."""
    inner = -lorentz_product(u, v)
    # Clamp to 1.0 to avoid NaN in acosh due to floating point inaccuracies
    arg = max(inner, 1.0)
    return math.acosh(arg)


def lorentz_to_poincare(x: Sequence[float]) -> List[float]:
    """Converts a point from the Lorentz model (Hyperboloid) to the Poincaré Ball model (129 -> 128)."""
    if not x:
        return []
    x0 = x[0]
    denom = max(1.0 + x0, 1e-12)
    return [xi / denom for xi in x[1:]]


def poincare_to_lorentz(p: Sequence[float]) -> List[float]:
    """Converts a point from the Poincaré Ball model to the Lorentz model (128 -> 129)."""
    p_sq = sum(v * v for v in p)
    denom = max(1.0 - p_sq, 1e-12)
    
    x = [(1.0 + p_sq) / denom]
    for pi in p:
        x.append(2.0 * pi / denom)
    return x


def project_to_hyperboloid(v: Sequence[float]) -> List[float]:
    """Ensures a vector satisfies the Lorentz constraint -x0^2 + |x|^2 = -1 (stabilization)."""
    if not v:
        return []
    res = list(v)
    spatial_norm_sq = sum(x * x for x in res[1:])
    res[0] = math.sqrt(1.0 + spatial_norm_sq)
    return res


# ==========================================
# Cognitive Math SDK (Spatial AI Engine)
# ==========================================
# ... rest of the file ...

def local_entropy(candidate: Sequence[float], neighbors: Sequence[Sequence[float]], c: float = 1.0) -> float:
    """
    Calculates the spatial entropy (dispersion) of a `candidate` vector relative to its `neighbors`.
    Used to track LLM hallucinations (Task 2.3.1).
    Returns a value in [0, 1) where values approaching 1 imply high chaos (hallucination).
    """
    if not neighbors:
        return 1.0  # Infinite entropy without neighbors
    total_deviation = 0.0
    for neighbor in neighbors:
        diff = log_map(candidate, neighbor, c=c)
        total_deviation += math.sqrt(max(_norm_sq(diff), 0.0))
    mean_deviation = total_deviation / len(neighbors)
    # Logarithmic compression mapping deviation to [0, 1)
    return 1.0 - math.exp(-mean_deviation)


def lyapunov_convergence(trajectory: Sequence[Sequence[float]], c: float = 1.0) -> float:
    """
    Evaluates if a trajectory of vectors (e.g. Chain of Thought) converges to an attractor.
    Calculates the average energy derivative (Lyapunov function derivative).
    Negative values indicate convergence (stable), positive indicate divergence (chaos/hallucination).
    """
    if len(trajectory) < 3:
        raise ValueError("Need at least 3 points to evaluate convergence trend")
    # Attractor is approximated by Fréchet mean of the trajectory
    attractor = frechet_mean(trajectory, c=c, max_iter=32, tol=1e-6)
    v_diff_sum = 0.0
    for i in range(len(trajectory) - 1):
        v_t0 = math.sqrt(max(_norm_sq(log_map(attractor, trajectory[i], c=c)), 0.0))
        v_t1 = math.sqrt(max(_norm_sq(log_map(attractor, trajectory[i + 1], c=c)), 0.0))
        v_diff_sum += (v_t1 - v_t0)
    
    return v_diff_sum / (len(trajectory) - 1)


def koopman_extrapolate(past: Sequence[float], current: Sequence[float], steps: float, c: float = 1.0) -> List[float]:
    """
    Extrapolates the trajectory in linear space (Koopman linearization) by tracking the 
    shift vector from `past` to `current` and projecting it forward.
    """
    velocity_at_past = log_map(past, current, c=c)
    velocity_at_current = parallel_transport(past, current, velocity_at_past, c=c)
    future_velocity = [v * steps for v in velocity_at_current]
    return exp_map(current, future_velocity, c=c)


def context_resonance(thought: Sequence[float], global_context: Sequence[float], resonance_factor: float, c: float = 1.0) -> List[float]:
    """
    Resonates a thought vector towards a global context vector (Phase-Locked Loop context synchronization).
    Pulls the thought towards the context along the geodesic by `resonance_factor` [0, 1].
    """
    pull_dir = log_map(thought, global_context, c=c)
    factor = max(0.0, min(1.0, resonance_factor))
    applied_pull = [v * factor for v in pull_dir]
    return exp_map(thought, applied_pull, c=c)


# ==========================================
# Private Projection / Mathematical Obfuscation (ZK-Privacy)
# ==========================================

def generate_orthogonal_matrix(dimension: int, seed_bytes: bytes):
    """
    Generates a deterministic, seed-based orthogonal matrix of shape (dimension, dimension).
    Preserves Euclidean distance and inner products (L2 / Cosine).
    """
    import numpy as np
    import hashlib
    seed_int = int.from_bytes(hashlib.sha256(seed_bytes).digest()[:8], 'big')
    rng = np.random.default_rng(seed_int)
    H = rng.standard_normal((dimension, dimension))
    Q, R = np.linalg.qr(H)
    d = np.diag(R)
    # Avoid division by zero
    d_abs = np.abs(d)
    d_abs[d_abs < 1e-15] = 1e-15
    ph = d / d_abs
    Q = Q * ph
    return Q


def generate_lorentz_matrix(dimension: int, seed_bytes: bytes):
    """
    Generates a deterministic, seed-based Lorentz transformation matrix of shape (dimension, dimension).
    Preserves Minkowski inner product (Lorentz distance / Hyperbolic metric).
    """
    import numpy as np
    import hashlib
    if dimension < 2:
        raise ValueError("Lorentz space dimension must be >= 2")
    d = dimension - 1
    seed_int = int.from_bytes(hashlib.sha256(seed_bytes).digest()[:8], 'big')
    rng = np.random.default_rng(seed_int)
    
    # 1. Generate random spatial rotation R
    H = rng.standard_normal((d, d))
    R, R_r = np.linalg.qr(H)
    d_diag = np.diag(R_r)
    d_diag_abs = np.abs(d_diag)
    d_diag_abs[d_diag_abs < 1e-15] = 1e-15
    ph = d_diag / d_diag_abs
    R = R * ph
    
    Lambda_R = np.eye(dimension)
    Lambda_R[1:, 1:] = R
    
    # 2. Generate random boost vector beta in R^d with ||beta|| in [0.1, 0.8]
    beta_dir = rng.standard_normal(d)
    norm = np.linalg.norm(beta_dir)
    if norm < 1e-9:
        beta_dir = np.zeros(d)
        beta_dir[0] = 1.0
        norm = 1.0
    beta_dir = beta_dir / norm
    v = 0.1 + 0.7 * rng.random()
    beta = v * beta_dir
    
    beta_sq = np.dot(beta, beta)
    gamma = 1.0 / np.sqrt(1.0 - beta_sq)
    
    Lambda_B = np.zeros((dimension, dimension))
    Lambda_B[0, 0] = gamma
    Lambda_B[0, 1:] = -gamma * beta
    Lambda_B[1:, 0] = -gamma * beta
    
    factor = (gamma - 1.0) / beta_sq
    spatial_part = np.eye(d) + factor * np.outer(beta, beta)
    Lambda_B[1:, 1:] = spatial_part
    
    return np.matmul(Lambda_R, Lambda_B)


def project_vector(v: Sequence[float], projection_matrix) -> List[float]:
    """Projects a vector using the provided projection matrix (v' = v * P)."""
    import numpy as np
    v_arr = np.array(v, dtype=np.float64)
    v_projected = np.matmul(v_arr, projection_matrix)
    return v_projected.tolist()


def inject_anisotropic_noise(vector: Sequence[float], noise_seed: bytes, sigma: float = 0.02) -> List[float]:
    """
    Injects deterministic, anisotropic noise calibrated to the vector norm.
    Protects against Manifold Alignment Attacks (Manifold Reconstruction).
    """
    if sigma <= 0.0:
        return list(vector)
    import numpy as np
    import hashlib
    vec_arr = np.array(vector, dtype=np.float64)
    # Generate deterministic seed from the vector data + noise_seed
    vec_bytes = str(vector).encode('utf-8')
    vec_hash = hashlib.sha256(noise_seed + vec_bytes).digest()
    seed_int = int.from_bytes(vec_hash[:8], 'big')
    rng = np.random.default_rng(seed_int)
    
    noise = rng.normal(0, sigma, len(vector))
    noisy = vec_arr + noise
    
    # Re-normalize to original norm to preserve cosine / lorentz distance metrics
    orig_norm = np.linalg.norm(vec_arr)
    noisy_norm = np.linalg.norm(noisy)
    if noisy_norm > 1e-15:
        noisy = (noisy / noisy_norm) * orig_norm
        
    return noisy.tolist()

File: sdks/dart/lib/src/math.dart
import 'dart:math';
import 'dart:typed_data';
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// HyperspaceDB Spatial and Cognitive Math SDK
/// Provides hyperbolic math functions and Cognitive AI metrics for solving LLM hallucinations.

double dot(List<double> a, List<double> b) {
  double sum = 0.0;
  for (int i = 0; i < a.length; i++) sum += a[i] * b[i];
  return sum;
}

double normSq(List<double> v) => dot(v, v);

double norm(List<double> v) => sqrt(max(normSq(v), 0.0));

List<double> projectToBall(List<double> x, double c) {
  double n = norm(x);
  double maxN = (1.0 / sqrt(c)) - 1e-9;
  if (n <= maxN || n <= 1e-15) return List.from(x);
  double s = maxN / n;
  return x.map((v) => v * s).toList();
}

List<double> mobiusAdd(List<double> x, List<double> y, [double c = 1.0]) {
  if (x.length != y.length) throw ArgumentError("Dimension mismatch");
  if (c <= 0.0) throw ArgumentError("Curvature c must be > 0");
  double xy = dot(x, y);
  double x2 = normSq(x);
  double y2 = normSq(y);
  double numLeft = 1.0 + 2.0 * c * xy + c * y2;
  double numRight = 1.0 - c * x2;
  double den = 1.0 + 2.0 * c * xy + c * c * x2 * y2;
  if (den.abs() < 1e-15) throw StateError("Möbius addition denominator too close to zero");
  List<double> res = [];
  for (int i = 0; i < x.length; i++) {
    res.add((numLeft * x[i] + numRight * y[i]) / den);
  }
  return res;
}

List<double> expMap(List<double> x, List<double> v, [double c = 1.0]) {
  if (x.length != v.length) throw ArgumentError("Dimension mismatch");
  if (c <= 0.0) throw ArgumentError("Curvature c must be > 0");
  double x2 = normSq(x);
  double vNorm = norm(v);
  if (vNorm < 1e-15) return List.from(x);
  double lambdaX = 2.0 / max(1.0 - c * x2, 1e-15);
  double scale = tanh(sqrt(c) * lambdaX * vNorm / 2.0) / (sqrt(c) * vNorm);
  List<double> step = v.map((vi) => scale * vi).toList();
  return mobiusAdd(x, step, c);
}

List<double> logMap(List<double> x, List<double> y, [double c = 1.0]) {
  if (x.length != y.length) throw ArgumentError("Dimension mismatch");
  if (c <= 0.0) throw ArgumentError("Curvature c must be > 0");
  List<double> negX = x.map((xi) => -xi).toList();
  List<double> delta = mobiusAdd(negX, y, c);
  double deltaNorm = norm(delta);
  if (deltaNorm < 1e-15) return List.filled(x.length, 0.0);
  double x2 = normSq(x);
  double lambdaX = 2.0 / max(1.0 - c * x2, 1e-15);
  double factor = (2.0 / (lambdaX * sqrt(c))) * atanh(min(sqrt(c) * deltaNorm, 1.0 - 1e-15));
  return delta.map((di) => factor * di / deltaNorm).toList();
}

List<double> gyro(List<double> u, List<double> v, List<double> w, [double c = 1.0]) {
  List<double> uv = mobiusAdd(u, v, c);
  List<double> vw = mobiusAdd(v, w, c);
  List<double> left = mobiusAdd(u, vw, c);
  List<double> negUv = uv.map((z) => -z).toList();
  return mobiusAdd(negUv, left, c);
}

List<double> parallelTransport(List<double> x, List<double> y, List<double> v, [double c = 1.0]) {
  if (x.length != y.length || x.length != v.length) throw ArgumentError("Dimension mismatch");
  if (c <= 0.0) throw ArgumentError("Curvature c must be > 0");
  List<double> negX = x.map((xi) => -xi).toList();
  List<double> gyr = gyro(y, negX, v, c);
  double lambdaX = 2.0 / max(1.0 - c * normSq(x), 1e-15);
  double lambdaY = 2.0 / max(1.0 - c * normSq(y), 1e-15);
  double scale = lambdaX / lambdaY;
  return gyr.map((gi) => scale * gi).toList();
}

List<double> frechetMean(List<List<double>> points, {double c = 1.0, int maxIter = 32, double tol = 1e-6}) {
  if (points.isEmpty) throw ArgumentError("Points set cannot be empty");
  if (c <= 0.0) throw ArgumentError("Curvature c must be > 0");
  int dim = points[0].length;
  List<double> mu = projectToBall(points[0], c);
  for (int iter = 0; iter < max(1, maxIter); iter++) {
    List<double> grad = List.filled(dim, 0.0);
    for (var p in points) {
      List<double> lg = logMap(mu, p, c);
      for (int i = 0; i < dim; i++) grad[i] += lg[i];
    }
    double inv = 1.0 / points.length;
    for (int i = 0; i < dim; i++) grad[i] *= inv;
    double gNorm = norm(grad);
    if (gNorm <= max(tol, 1e-15)) break;
    mu = expMap(mu, grad, c);
    mu = projectToBall(mu, c);
  }
  return mu;
}

double poincareDist(List<double> u, List<double> v, [double c = 1.0]) {
  double u2 = normSq(u);
  double v2 = normSq(v);
  double diff2 = 0.0;
  for (int i = 0; i < u.length; i++) diff2 += pow(u[i] - v[i], 2);
  double denom = (1.0 - c * u2) * (1.0 - c * v2);
  double arg = 1.0 + 2.0 * c * diff2 / max(denom, 1e-15);
  return acosh(arg);
}

// ==========================================
// Lorentz Model Math (Hyperboloid)
// ==========================================

double lorentzProduct(List<double> u, List<double> v) {
  if (u.isEmpty || v.isEmpty) return 0.0;
  double product = -u[0] * v[0];
  for (int i = 1; i < u.length; i++) product += u[i] * v[i];
  return product;
}

double lorentzDist(List<double> u, List<double> v) {
  double inner = -lorentzProduct(u, v);
  return acosh(max(inner, 1.0));
}

List<double> lorentzToPoincare(List<double> x) {
  if (x.isEmpty) return [];
  double x0 = x[0];
  double denom = max(1.0 + x0, 1e-12);
  return x.skip(1).map((xi) => xi / denom).toList();
}

List<double> poincareToLorentz(List<double> p) {
  double p2 = normSq(p);
  double denom = max(1.0 - p2, 1e-12);
  List<double> x = [(1.0 + p2) / denom];
  for (var pi in p) {
    x.add(2.0 * pi / denom);
  }
  return x;
}

List<double> projectToHyperboloid(List<double> v) {
  if (v.isEmpty) return [];
  List<double> res = List.from(v);
  double spatialNormSq = 0.0;
  for (int i = 1; i < res.length; i++) spatialNormSq += res[i] * res[i];
  res[0] = sqrt(1.0 + spatialNormSq);
  return res;
}

// Helper Hyperbolic Math Utilities
double tanh(double x) {
  double exp2x = exp(2.0 * x);
  return (exp2x - 1.0) / (exp2x + 1.0);
}

double atanh(double x) {
  return 0.5 * log((1.0 + x) / (1.0 - x));
}

double acosh(double x) {
  return log(x + sqrt(x * x - 1.0));
}

// ==========================================
// Cognitive Math SDK (Spatial AI Engine)
// ==========================================

double localEntropy(List<double> candidate, List<List<double>> neighbors, {double c = 1.0}) {
  if (neighbors.isEmpty) return 1.0;
  double totalDeviation = 0.0;
  for (var neighbor in neighbors) {
    List<double> diff = logMap(candidate, neighbor, c);
    totalDeviation += norm(diff);
  }
  double meanDeviation = totalDeviation / neighbors.length;
  return 1.0 - exp(-meanDeviation);
}

double lyapunovConvergence(List<List<double>> trajectory, {double c = 1.0}) {
  if (trajectory.length < 3) throw ArgumentError("Need at least 3 points");
  List<double> attractor = frechetMean(trajectory, c: c, maxIter: 32, tol: 1e-6);
  double vDiffSum = 0.0;
  for (int i = 0; i < trajectory.length - 1; i++) {
    double vt0 = norm(logMap(attractor, trajectory[i], c));
    double vt1 = norm(logMap(attractor, trajectory[i + 1], c));
    vDiffSum += (vt1 - vt0);
  }
  return vDiffSum / (trajectory.length - 1);
}

List<double> koopmanExtrapolate(List<double> past, List<double> current, double steps, {double c = 1.0}) {
  List<double> velocityAtPast = logMap(past, current, c);
  List<double> velocityAtCurrent = parallelTransport(past, current, velocityAtPast, c);
  List<double> futureVelocity = velocityAtCurrent.map((v) => v * steps).toList();
  return expMap(current, futureVelocity, c);
}

List<double> contextResonance(List<double> thought, List<double> globalContext, double resonanceFactor, {double c = 1.0}) {
  List<double> pullDir = logMap(thought, globalContext, c);
  double factor = max(0.0, min(1.0, resonanceFactor));
  List<double> appliedPull = pullDir.map((v) => v * factor).toList();
  return expMap(thought, appliedPull, c);
}

List<List<double>> generateOrthogonalMatrix(int dimension, List<int> seedBytes) {
  final matrix = List.generate(dimension, (_) => List<double>.filled(dimension, 0.0));
  var currentHash = sha256.convert(seedBytes).bytes;
  var hashOffset = 0;
  final byteData = ByteData(4);

  for (int i = 0; i < dimension; i++) {
    for (int j = 0; j < dimension; j++) {
      if (hashOffset >= 32) {
        currentHash = sha256.convert(currentHash).bytes;
        hashOffset = 0;
      }
      byteData.setUint8(0, currentHash[hashOffset]);
      byteData.setUint8(1, currentHash[hashOffset + 1]);
      byteData.setUint8(2, currentHash[hashOffset + 2]);
      byteData.setUint8(3, currentHash[hashOffset + 3]);
      final valInt = byteData.getUint32(0, Endian.little);
      final val = (valInt / 4294967296.0) * 2.0 - 1.0;
      matrix[i][j] = val;
      hashOffset += 4;
    }
  }

  // Gram-Schmidt with Reorthogonalization
  for (int i = 0; i < dimension; i++) {
    var v = List<double>.from(matrix[i]);
    for (int j = 0; j < i; j++) {
      final u = matrix[j];
      var uDotV = dot(u, v);
      for (int k = 0; k < dimension; k++) {
        v[k] -= uDotV * u[k];
      }
      uDotV = dot(u, v);
      for (int k = 0; k < dimension; k++) {
        v[k] -= uDotV * u[k];
      }
    }
    final vNorm = norm(v);
    if (vNorm > 1e-15) {
      for (int k = 0; k < dimension; k++) {
        matrix[i][k] = v[k] / vNorm;
      }
    } else {
      for (int k = 0; k < dimension; k++) {
        matrix[i][k] = 0.0;
      }
      matrix[i][i] = 1.0;
    }
  }

  return matrix;
}

List<List<double>> generateLorentzMatrix(int dimension, List<int> seedBytes) {
  final d = dimension - 1;

  final rSeed = sha256.convert([...seedBytes, ...utf8.encode("spatial")]).bytes;
  final R = generateOrthogonalMatrix(d, rSeed);

  final bSeed = sha256.convert([...seedBytes, ...utf8.encode("boost")]).bytes;
  var beta = List<double>.filled(d, 0.0);
  var currentHash = bSeed;
  var hashOffset = 0;
  final byteData = ByteData(4);

  for (int i = 0; i < d; i++) {
    if (hashOffset >= 32) {
      currentHash = sha256.convert(currentHash).bytes;
      hashOffset = 0;
    }
    byteData.setUint8(0, currentHash[hashOffset]);
    byteData.setUint8(1, currentHash[hashOffset + 1]);
    byteData.setUint8(2, currentHash[hashOffset + 2]);
    byteData.setUint8(3, currentHash[hashOffset + 3]);
    final valInt = byteData.getUint32(0, Endian.little);
    final val = (valInt / 4294967296.0) * 2.0 - 1.0;
    beta[i] = val;
    hashOffset += 4;
  }

  final betaNorm = norm(beta);
  if (betaNorm > 1e-15) {
    for (int i = 0; i < d; i++) {
      beta[i] = (beta[i] / betaNorm) * 0.1;
    }
  } else {
    for (int i = 0; i < d; i++) {
      beta[i] = 0.0;
    }
    beta[0] = 0.1;
  }

  final betaSq = dot(beta, beta);
  final gamma = 1.0 / sqrt(1.0 - betaSq);

  final Lambda_B = List.generate(dimension, (_) => List<double>.filled(dimension, 0.0));
  Lambda_B[0][0] = gamma;
  for (int j = 1; j < dimension; j++) {
    Lambda_B[0][j] = -gamma * beta[j - 1];
    Lambda_B[j][0] = -gamma * beta[j - 1];
  }

  final factor = (gamma - 1.0) / betaSq;
  for (int i = 1; i < dimension; i++) {
    for (int j = 1; j < dimension; j++) {
      final delta = i == j ? 1.0 : 0.0;
      Lambda_B[i][j] = delta + factor * beta[i - 1] * beta[j - 1];
    }
  }

  final Lambda_R = List.generate(dimension, (_) => List<double>.filled(dimension, 0.0));
  Lambda_R[0][0] = 1.0;
  for (int i = 1; i < dimension; i++) {
    for (int j = 1; j < dimension; j++) {
      Lambda_R[i][j] = R[i - 1][j - 1];
    }
  }

  final Lambda = List.generate(dimension, (_) => List<double>.filled(dimension, 0.0));
  for (int i = 0; i < dimension; i++) {
    for (int j = 0; j < dimension; j++) {
      double sum = 0.0;
      for (int k = 0; k < dimension; k++) {
        sum += Lambda_B[i][k] * Lambda_R[k][j];
      }
      Lambda[i][j] = sum;
    }
  }

  return Lambda;
}

List<double> projectVector(List<double> v, List<List<double>> matrix) {
  final dim = v.length;
  final projected = List<double>.filled(dim, 0.0);
  for (int j = 0; j < dim; j++) {
    double sum = 0.0;
    for (int i = 0; i < dim; i++) {
      sum += v[i] * matrix[i][j];
    }
    projected[j] = sum;
  }
  return projected;
}

List<double> injectAnisotropicNoise(List<double> v, List<int> seedBytes, double sigma) {
  if (v.isEmpty || sigma <= 0.0) return List.from(v);
  final dim = v.length;
  final noise = List<double>.filled(dim, 0.0);
  var currentHash = sha256.convert(seedBytes).bytes;
  var hashOffset = 0;
  final byteData = ByteData(4);

  for (int i = 0; i < dim; i++) {
    if (hashOffset >= 32) {
      currentHash = sha256.convert(currentHash).bytes;
      hashOffset = 0;
    }
    byteData.setUint8(0, currentHash[hashOffset]);
    byteData.setUint8(1, currentHash[hashOffset + 1]);
    byteData.setUint8(2, currentHash[hashOffset + 2]);
    byteData.setUint8(3, currentHash[hashOffset + 3]);
    final valInt = byteData.getUint32(0, Endian.little);
    final val = (valInt / 4294967296.0) * 2.0 - 1.0;
    noise[i] = val;
    hashOffset += 4;
  }

  final vNorm = norm(v);
  if (vNorm > 1e-15) {
    final projLength = dot(noise, v) / (vNorm * vNorm);
    for (int i = 0; i < dim; i++) {
      noise[i] -= projLength * v[i];
    }
  }

  final noiseNorm = norm(noise);
  if (noiseNorm > 1e-15) {
    final scale = sigma * vNorm / noiseNorm;
    return List.generate(dim, (i) => v[i] + noise[i] * scale);
  } else {
    return List.from(v);
  }
}
